src/jdk.compiler/share/classes/com/sun/tools/javac/parser/DocCommentParser.java

```diff
@@ -1124,6 +1308,375 @@ protected Name readSystemPropertyName() {
         return names.fromChars(buf, pos, bp - pos);
     }
 
+    /**
+     * A class to encapsulate the work to detect the boundaries of character
+     * sequences of literal text, and to skip over such sequences, without
+     * analyzing the content.
+     *
+     * The two primary factors are the current level of indentation,
+     * used when checking for indented code blocks, and a serial number
+     * that is incremented when a line is encountered that is not a
+     * continuation of the previous block.
+     */
+    class Markdown {
+        /**
+         * The current "line kind".
+         */
+        private LineKind lineKind = LineKind.OTHER;
+
+        /**
+         * A serial number for the current "block".
+         * It is incremented whenever a line is encountered that indicates
+         * the beginning of a new block, such that any potential code span
+         * should be terminated as "not a span".
+         */
+        private int blockId = 0;
+
+        /**
+         * The current indentation used to determine if a line is part of
+         * an indented code block. Initially zero, it is increased while
+         * analysing bulleted or ordered list items,
+         */
+        private int currIndent = 0;
+
+        /**
+         * A stack of the indentation values for nested lists.
+         * Initially empty, entries are pushed when a top-level or
+         * nested list item is encountered, and removed when a
+         * less indented line is encountered.
+         */
+        private final Deque<Integer> indentStack = new ArrayDeque<>();
+
+        /**
+         * Updates the state after a newline has been read.
+         * There are two primary goals.
+         *
+         * 1. Determine the indentation used to identify indented code blocks.
+         * 2. Determine if the current line should be considered a continuation
+         *    of the preceding block, or the beginning of a new block. This is
+         *    used when determining the end (if any) of a potential code span.
+         */
+        void update() {
+            var prevLineKind = lineKind;
+            int indent = readIndent(0);
+            if (ch == '\r' || ch == '\n') {
+                lineKind = LineKind.BLANK;
+                blockId++;
+            } else if (indent >= currIndent + 4 && !isParagraph(prevLineKind)) {
+                if (prevLineKind != LineKind.INDENTED_CODE_BLOCK) {
+                    blockId++;
+                }
+                lineKind = LineKind.INDENTED_CODE_BLOCK;
+            } else {
+                while (indent < currIndent) {
+                    currIndent = indentStack.pop();
+                }
```

Comment:
This new Markdown class looks very good!

I think in this place we also need to check for indented code blocks, since we reduced `currIndent` and may have 4 or more character indentation compared to the new `currIndent` value. 

My tentative fix is to add the following code here, but maybe there's a more elegant solution by restructuring the method to first check for `indent < currIndent` and then for new indented code blocks.

```
                if (indent >= currIndent + 4 && !isParagraph(prevLineKind)) {
                    blockId++;
                    lineKind = LineKind.INDENTED_CODE_BLOCK;
                    return;
                }
```

The following could be added to `TestMarkdownCodeBlocks.java` to test this case:

```
            POST_LIST_INDENT(
                    """
                         1.  list item
                         
                             second paragraph
                       
                            {@code CODE}
                            @Anno

                        end""",
                    """
                        <ol>
                        <li>
                        <p>list item</p>
                        <p>second paragraph</p>
                        </li>
                        </ol>
                        <pre><code>{@code CODE}
                        @Anno
                        </code></pre>
                        <p>end</p>"""),
```

Reply:
Thanks.  I will investigate your hint to look for a more elegant solution.  Thanks for the new test case.

Reply:
>I think in this place we also need to check for indented code blocks, since we reduced currIndent and may have 4 or more character indentation compared to the new currIndent value.

This seems wrong. Surely, we should not reduce the indentation and then for the same line in the comment check for an extra-indented line. Or else I am missing something.

That being said, thanks for the test case. I will investigate it.

Reply:
Update: I see and understand your test case.

Reply:
Update: I found that more elegant solution you hinted at, by refactoring the method from nested `if` statements to a series of checks with early returns.

The test case you provided now works as expected.